docs/dev/Lua API.rst

```diff
@@ -5012,6 +5012,39 @@ The widget implements:
 
   Same as with an ordinary list.
 
+TabBar class
+------------
+
+This widget implements a set of one or more tabs to allow navigation between groups of content.
+
+:labels: A table of strings; entry representing the label text for a single tab. The order of the entries
+         determines the order the tabs will appear in.
+:get_cur_page: Returns the current "page". This function does not have a default implementation; you must provide
+               an implementation that returns the current value of whichever variable your script uses to keep track of the
+               current "page" (this does not need to relate to an actual Pages widget).
+:on_select: Callback executed when a tab is selected. It receives the selected tab index as an argument. Your implementation
+            should likely update the value of whichever variable your script uses to keep track of the current page.
+:active_tab_pens: A table of pens used to render active tabs. See the default implementation in widgets.lua for an example
+                  of how to construct the table. Leave unspecified to use the default pens.
+:inactive_tab_pens: A table of pens used to render inactive tabs. See the default implementation in widgets.lua for an example
+                    of how to construct the table. Leave unspecified to use the default pens.
+:get_pens: A function used to determine which pens should be used to render a tab. Receives the index of the tab as the first
+           argument and the TabBar widget itself as the second. The default implementation, which will handle most situations,
+           returns ``self.active_tab_pens``, if ``self.get_cur_page() == idx``, otherwise returns ``self.inactive_tab_pens``.
+:switch_tab_key: Specifies a keybinding that can be used to switch to the next tab.
+
+Tab class
+---------
+
+This widget implements a single clickable tab and is the main component of the TabBar widget. Usage of the ``TabBar``
+widget does not require direct usage of ``Tab``.
+
+:id: The id of the tab.
+:label: The text displayed on the tab.
+:on_select: Callback executed when the tab is selected.
+:get_pens: A function that is used during ``Tab:onRenderBody`` to determine the pens that should be used for drawing. See the
+           usage of ``Tab`` in ``TabBar:init()`` for an example. See the default value of ``active_tab_pens`` or ``inactive_tab_pens``
+           in ``TabBar`` for an example of how to construct pens.
```

Comment:
This is why I didn't push the widgets straight into the library. The API as it stands just feels overly complicated and unencapsulated.

I think what I'd like to see is there not being any externally-visible Tab or TabBar class at all. Instead, widget.Pages can take a `labels` attribute, and if it is set, then it creates and manages the tabs itself. Is there a strong use case for these tabs *outside* of a Pages widget?

Reply:
I don't know if it's a strong use case but.. I'll describe how I'm currently using it outside of a Pages widget and see how that goes haha

In that units search/overview script I've been working on:

![image](https://user-images.githubusercontent.com/2104555/219174493-254844f5-203d-4fd8-9e4c-d5631171fc48.png)

I'm [not using a Pages widget](https://gist.github.com/robob27/1a30758877566def65f80a8ae3a22fc3#file-unit-search-lua-L1128-L1324). I make the top TabBar responsible for controlling a variable called `view_mode`, and the second TabBar responsible for a variable called `active_subviews`. Each time the view mode or subview mode is updated, I refresh the `units` table [with units that are valid for the current mode/submode](https://gist.github.com/robob27/1a30758877566def65f80a8ae3a22fc3#file-unit-search-lua-L567-L634) and then update the single shared `FilteredList` called `unit_list`. All of the hotkeylabels etc are shared by all views and just have their labels and callbacks updated as needed based on the view/subview mode.

It feels like if I were to use a Pages widget for this, I would end up having to duplicate a lot of widgets to achieve the same sort of layout. For my particular use case, it feels like I just want a pretty way to control the value of 2 variables rather than a way to control the visibility of grouped sets of widgets.

Reply:
That sounds legit. Essentially a CycleHotkeyLabel with all options visible and directly selectable.